File: src/_cpp/TowerDefenseBase/scenes/gamegs/GamePauseScene.cpp
//
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 27.09.14.
//
//
#include "GameGS.h"
#include "GamePauseScene.h"
#include "consts.h"
#include "MenuItemTextBG.h"
#include "ml/ScrollMenu.h"
#include "gameboard.h"
#include "ml/Audio/AudioEngine.h"
#include "ml/Audio/AudioMenu.h"
#include "resources.h"
#include "ml/Language.h"
#include "ScoreCounter.h"
#include "ml/SmartScene.h"
#include "ShopLayer.h"
#include "ScoreLayer.h"
#include "configuration.h"
#include "MenuItemWithText.h"
#include "UserData.h"

NS_CC_BEGIN;

//const int zOrderBG( -10 );
//const int zOrderShadow( 0 );
//const int zOrderLabels( 10 );
//const int zOrderMenu( 20 );
//const int zOrderOptins( 30 );

float kFadeDuration( 0.2f );

class DialogRestartGame : public LayerExt
{
	DECLARE_BUILDER( DialogRestartGame );
	bool init()
	{
		LayerExt::init();
		load( "ini/maings/dialogclosegame.xml" );
		return true;
	}
	ccMenuCallback get_callback_by_description( const std::string & name )
	{
		if( name == "yes" ) return std::bind( []( Ref* ){ Director::getInstance()->end(); }, std::placeholders::_1 );
		if( name == "no" ) return std::bind( [this]( Ref* ){ removeFromParent(); }, std::placeholders::_1 );
		return LayerExt::get_callback_by_description( name );
	}
};
DialogRestartGame::DialogRestartGame() {}
DialogRestartGame::~DialogRestartGame() {}

GamePauseLayer::GamePauseLayer()
: _scaleFactor( 1 )
{
};

bool GamePauseLayer::init( const std::string& path, bool showScores )
{
	do
	{
		Size desSize = Director::getInstance()->getOpenGLView()->getDesignResolutionSize();
		CC_BREAK_IF( !Menu::init() );
		setPosition( Point( desSize / 2 ) );
		setContentSize( Size::ZERO );
		setKeyboardEnabled( true );
		_showScores = showScores;

		load( path );

		_resume = getChildByName<MenuItem*>( "resume" );
		_restart = getChildByName<MenuItem*>( "restart" );
		_quit = getChildByName<MenuItem*>( "quit" );
		_store = getChildByName<MenuItem*>( "store" );
		_music_on = getChildByName<MenuItem*>( "music_on" );
		_music_off = getChildByName<MenuItem*>( "music_off" );
		_sound_on = getChildByName<MenuItem*>( "sound_on" );
		_sound_off = getChildByName<MenuItem*>( "sound_off" );
        
//        auto bL = getChildByName("restart");
//        if (bL) {
//            ((MenuItemImageWithText *)bL)->setFont2(kFontStroke);
//            ((MenuItemImageWithText *)bL)->setText2(WORD("restart"));
//            auto text2 = (Label *)bL->getChildByName("normal")->getChildByName("text2");
//            if (text2) {
//                text2->setPositionY( 58.f );
//                text2->setScale( 0.45f );
//            }
//        }

		auto cost = _restart?_restart->getChildByName<Label*>( "normal" ) : nullptr;
        cost = cost?cost->getChildByName<Label*>( "cost" ) : nullptr;

		if( cost )
		{
			int index = GameGS::getInstance()->getGameBoard().getCurrentLevelIndex();
			GameMode mode = GameGS::getInstance()->getGameBoard().getGameMode();
			int value = LevelParams::shared().getFuel( index, mode == GameMode::hard );
			cost->setString(intToStr(value));
		}


		auto bg = getChildByName( "bg" );
		if( bg )
		{
			auto size = bg->getContentSize();
			float sx = std::min<float>( 1, desSize.width / size.width );
			float sy = std::min<float>( 1, desSize.height / size.height );
			float s = std::min<float>( sx, sy );
			s = std::min<float>( 1, s );
			_scaleFactor = s;
		}
        
        if (k::configuration::useFuel) {
            auto restart_text = this->getChildByPath("restart/normal/restart_text");
            if (restart_text) {
                restart_text->setPositionY(restart_text->getPositionY() + 14.f);
            }
            auto restart_fuel_icon = this->getChildByPath("restart/normal/icon");
            if (restart_fuel_icon) restart_fuel_icon->setVisible(true);
            auto restart_fuel_cost = this->getChildByPath("restart/normal/cost");
            if (restart_fuel_cost) restart_fuel_cost->setVisible(true);
        }

		setOpacity( 0 );
		fadeenter();
		checkAudio();
		checkFullscreen();

		return true;
	}
	while( false );
	return false;
}

GamePauseLayer::~GamePauseLayer()
{
	removeAllChildrenWithCleanup( true );
	//Director::getInstance()->getTextureCache()->removeUnusedTextures();
};

void GamePauseLayer::onEnter()
{
	Menu::onEnter();
	setKeyboardEnabled( true );
	AudioEngine::shared().playEffect( kSoundGamePauseOn );

	if( _showScores )
	{
		auto scene = Director::getInstance()->getRunningScene();
		auto scores = scene->getChildByName( "scorelayer" );
		if( !scores )
		{
			auto scores = ScoreLayer::create();
			scene->addChild( scores, 999 );
		}
	}
	if( _resume ) _resume->setEnabled( true );
	if( _restart ) _restart->setEnabled( true );
}

void GamePauseLayer::onKeyReleased( EventKeyboard::KeyCode keyCode, Event* event )
{
	if( keyCode == EventKeyboard::KeyCode::KEY_BACK )
		cb_resume( nullptr );
}

ccMenuCallback GamePauseLayer::get_callback_by_description( const std::string & name )
{
	if( name == "resume" ) return CC_CALLBACK_1( GamePauseLayer::cb_resume, this );
	if( name == "restart" ) return CC_CALLBACK_1( GamePauseLayer::cb_restart, this );
	if( name == "quit" ) return CC_CALLBACK_1( GamePauseLayer::cb_exit, this );
	if( name == "music_on" ) return CC_CALLBACK_1( GamePauseLayer::cb_music, this, false );
	if( name == "music_off" ) return CC_CALLBACK_1( GamePauseLayer::cb_music, this, true );
	if( name == "sound_on" ) return CC_CALLBACK_1( GamePauseLayer::cb_sound, this, false );
	if( name == "sound_off" ) return CC_CALLBACK_1( GamePauseLayer::cb_sound, this, true );
	if( name == "fullscreen" )
	{
		auto cb = [this]( Ref*sender )
		{
			auto fullscreen = !UserData::shared().get_bool( "fullscreen", true );
			UserData::shared().write( "fullscreen", fullscreen );
			checkFullscreen();

			auto layer = DialogRestartGame::create();
			auto scene = static_cast<SmartScene*>(getScene());
			scene->pushLayer( layer, true );
		};
		return std::bind( cb, std::placeholders::_1 );
	}
	return nullptr;
}

void GamePauseLayer::cb_resume( Ref*sender )
{
	auto func = CallFunc::create( std::bind( &GamePauseLayer::gameresume, this ) );
	runAction( Sequence::create( DelayTime::create( kFadeDuration ), func, nullptr ) );
	fadeexit();
}

void GamePauseLayer::cb_restart( Ref*sender )
{
	int index = GameGS::getInstance( )->getGameBoard( ).getCurrentLevelIndex( );
	GameMode mode = GameGS::getInstance()->getGameBoard().getGameMode();
	int cost = LevelParams::shared().getFuel( index, GameMode::hard == mode );
	int fuel = ScoreCounter::shared( ).getMoney( kScoreFuel );
	if( cost <= fuel )
	{
		auto delay = DelayTime::create( kFadeDuration );
		auto func = CallFunc::create( std::bind( &GamePauseLayer::restart, this ) );
		runAction( Sequence::create( delay, func, nullptr ) );
		fadeexit();
	}
	else
	{
#if PC != 1
		auto shop = ShopLayer::create(k::configuration::useFreeFuel, false, false, false);
		if( shop )
		{
			auto scene = Director::getInstance()->getRunningScene();
			auto smartscene = dynamic_cast<SmartScene*>(scene);
			assert(smartscene);
			smartscene->pushLayer(shop, true);
		}
#endif
	}
	if( _resume ) _resume->setEnabled( false );
	if( _restart ) _restart->setEnabled( false );
}

void GamePauseLayer::cb_exit( Ref*sender )
{
	auto delay = DelayTime::create( kFadeDuration );
	auto func = CallFunc::create( std::bind( &GamePauseLayer::exit, this ) );
	runAction( Sequence::create( delay, func, nullptr ) );
	fadeexit();

	if( _resume ) _resume->setEnabled( false );
	if( _restart ) _restart->setEnabled( false );
}

void GamePauseLayer::cb_sound( Ref*sender, bool enabled )
{
	AudioEngine::shared().soundEnabled(enabled);
	checkAudio();
}

void GamePauseLayer::cb_music( Ref*sender, bool enabled )
{
	AudioEngine::shared().musicEnabled(enabled);
	checkAudio();
}

void GamePauseLayer::checkAudio()
{
	bool s = AudioEngine::shared().isSoundEnabled();
	bool m = AudioEngine::shared().isMusicEnabled();
	if( _sound_off )_sound_off->setVisible( !s );
	if( _sound_on  )_sound_on->setVisible( s );
	if( _music_off )_music_off->setVisible( !m );
	if( _music_on )_music_on->setVisible( m );
}

void GamePauseLayer::gameresume()
{
	AudioEngine::shared().resumeAllEffects();
	setEnabled( false );
	removeFromParent();
}

void GamePauseLayer::restart()
{
	setEnabled( false );
	removeFromParent();
	GameGS::restartLevel();
}

void GamePauseLayer::exit()
{
	setEnabled( false );
	Director::getInstance()->popScene();
	removeFromParent();
}

void GamePauseLayer::shop_did_closed()
{
	setEnabled( true );
}

void GamePauseLayer::fadeexit()
{
	auto scene = Director::getInstance()->getRunningScene();
	auto scores = scene->getChildByName("scorelayer");
	if(scores)
	{
		scores->removeFromParent();
	}

	runAction( ScaleTo::create( kFadeDuration, _scaleFactor * 1.2f ) );
	runAction( FadeTo::create( kFadeDuration, 0 ) );
}

void GamePauseLayer::fadeenter()
{
	setScale( _scaleFactor * 1.2f );
	runAction( ScaleTo::create( kFadeDuration, _scaleFactor * 1.f ) );
	runAction( FadeTo::create( kFadeDuration, 255 ) );
}

void GamePauseLayer::checkFullscreen()
{
	auto fullscreen = UserData::shared().get_bool( "fullscreen", true );
	auto item = getNodeByPath<mlMenuItem>( this, "fullscreen" );
	if( item )
	{
		auto on = item->getParamCollection().get( "on", "" );
		auto off = item->getParamCollection().get( "off", "" );
		item->setImageNormal( fullscreen ? on : off );
	}
}

NS_CC_END;

File: src/_eu/scenes/gamegs/GamePauseScene.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};

EU.GamePauseLayer = cc.Menu.extend(
{
    //bool init(   path, bool showScores = true );
    //Sprite m_shadow;
    /** @type {MenuItem} */ _resume : null ,
    /** @type {MenuItem} */ _restart : null ,
    /** @type {MenuItem} */ _quit : null,
    /** @type {MenuItem} */ _store : null,
    /** @type {MenuItem} */ _music_on : null,
    /** @type {MenuItem} */ _music_off : null,
    /** @type {MenuItem} */ _sound_on : null,
    /** @type {MenuItem} */ _sound_off : null,
    /** @type {bool} */ _showScores : null,
    /** @type {Number} */ _scaleFactor : 1.0,
    /** @type {Number} */ kFadeDuration : 0.2,

    DialogRestartGame : EU.LayerExt.extend(
    {
        ctor: function()
        {
            this._super();
            this.load_str( "ini/maings/dialogclosegame.xml" );
        },
        get_callback_by_description: function( name )
        {
            if( name == "yes" ) return cc.director.end.bind(this);
            if( name == "no" ) return this.removeFromParent.bind(this);
            return null;
        }
    }),

    ctor: function(  path, showScores )
    {
        if (showScores == null || showScores === undefined) showScores = true;

        var desSize = cc.view.getDesignResolutionSize();
        this._super();
        this.setPosition( Point( desSize / 2 ) );
        this.setContentSize( Size.ZERO );
        var self = this;
        cc.eventManager.addListener({
            event: cc.EventListener.KEYBOARD,
            onKeyReleased: function( keyCode, event )
            {
                if( keyCode == cc.KEY.back)
                    self.cb_resume( null );
            }
        }, this);
        this._showScores = showScores;

        this.load_str( path );

        this._resume = this.getChildByName( "resume" );
        this._restart = this.getChildByName( "restart" );
        this._quit = this.getChildByName( "quit" );
        this._store = this.getChildByName( "store" );
        this._music_on = this.getChildByName( "music_on" );
        this._music_off = this.getChildByName( "music_off" );
        this._sound_on = this.getChildByName( "sound_on" );
        this._sound_off = this.getChildByName( "sound_off" );

//        var bL = this.getChildByName("restart");
//        if (bL) {
//            ((MenuItemImageWithText *)bL).setFont2(kFontStroke);
//            ((MenuItemImageWithText *)bL).setText2(WORD("restart"));
//            var text2 = (Label *)bL.getChildByName("normal").getChildByName("text2");
//            if (text2) {
//                text2.setPositionY( 58.f );
//                text2.setScale( 0.45 );
//            }
//        }

        var cost = this._restart ? this._restart.getChildByName( "normal" ) : null;
        cost = cost ? cost.getChildByName( "cost" ) : null;

        if( cost )
        {
            var index = EU.GameGS.getInstance().getGameBoard().getCurrentLevelIndex();
            var mode = EU.GameGS.getInstance().getGameBoard().getGameMode();
            var value = EU.LevelParams.getFuel( index, mode == EU.GameMode.hard );
            cost.setString((value));
        }


        var bg = this.getChildByName( "bg" );
        if( bg )
        {
            var size = bg.getContentSize();
            var sx = Math.min( 1, desSize.width / size.width );
            var sy = Math.min( 1, desSize.height / size.height );
            var s = Math.min( sx, sy );
            s = Math.min( 1, s );
            this._scaleFactor = s;
        }

        if (EU.k.useFuel) {
            var restart_text = this.getChildByPath_str("restart/normal/restart_text");
            if (restart_text) {
                restart_text.setPositionY(restart_text.getPositionY() + 14);
            }
            var restart_fuel_icon = this.getChildByPath_str("restart/normal/icon");
            if (restart_fuel_icon) restart_fuel_icon.setVisible(true);
            var restart_fuel_cost = this.getChildByPath_str("restart/normal/cost");
            if (restart_fuel_cost) restart_fuel_cost.setVisible(true);
        }

        this.setOpacity( 0 );
        this.fadeenter();
        this.checkAudio();
        this.checkFullscreen();

    },

    onExit: function()
    {
        this.removeAllChildrenWithCleanup( true );
        //cc.director.getTextureCache().removeUnusedTextures();
        this._super();
    },

    onEnter: function()
    {
        this._super();
        //this.setKeyboardEnabled( true );
        EU.AudioEngine.playEffect( EU.kSoundGamePauseOn , false);

        if( this._showScores )
        {
            var scene = cc.director.getRunningScene();
            var scores = scene.getChildByName( "scorelayer" );
            if( !scores )
            {
                var scores = new EU.ScoreLayer();
                scene.addChild( scores, 999 );
            }
        }
        if( this._resume ) this._resume.setEnabled( true );
        if( this._restart ) this._restart.setEnabled( true );
    },

    get_callback_by_description: function( name )
    {
        if( name == "resume" ) return this.cb_resume.bind(this);
        if( name == "restart" ) return this.cb_restart.bind(this);
        if( name == "quit" ) return this.cb_exit.bind(this);
        if( name == "music_on" ) return this.cb_music.bind(this, false );
        if( name == "music_off" ) return this.cb_music.bind(this, true );
        if( name == "sound_on" ) return this.cb_sound.bind(this, false );
        if( name == "sound_off" ) return this.cb_sound.bind(this, true );
        if( name == "fullscreen" )
        {
            var cb = function()
            {
                var fullscreen = !EU.UserData.get_bool( "fullscreen", true );
                EU.UserData.write( "fullscreen", fullscreen );
                this.checkFullscreen();

                var layer = new this.DialogRestartGame();
                var scene = EU.Common.getSceneOfNode(this);
                scene.pushLayer( layer, true );
            };
            return cb.bind(this);
        }
        return null;
    },

    cb_resume: function()
    {
        var func = cc.callFunc( this.gameresume.bind( this ) );
        this.runAction( cc.sequence( cc.delayTime( this.kFadeDuration ), func ) );
        this.fadeexit();
    },

    cb_restart: function( sender )
    {
        var index = EU.GameGS.getInstance().getGameBoard( ).getCurrentLevelIndex( );
        var mode = EU.GameGS.getInstance().getGameBoard().getGameMode();
        var cost = EU.LevelParams.getFuel( index, EU.GameMode.hard == mode );
        var fuel = EU.ScoreCounter.getMoney( kScoreFuel );
        if( cost <= fuel )
        {
            var delay = cc.delayTime( this.kFadeDuration );
            var func = cc.callFunc( this.restart.bind( this ) );
            this.runAction( cc.sequence( delay, func ) );
            this.fadeexit();
        }
        else
        {
            if (EU.PC != 1) {
                var shop = new EU.ShopLayer(EU.k.useFreeFuel, false, false, false);
                if (shop) {
                    var scene = cc.director.getRunningScene();
                    var smartscene = scene;
                    EU.assert(smartscene.__SmartScene);
                    smartscene.pushLayer(shop, true);
                }
            }
        }
        if( this._resume ) this._resume.setEnabled( false );
        if( this._restart ) this._restart.setEnabled( false );
    },

    cb_exit: function( sender )
    {
        var delay = cc.delayTime( this.kFadeDuration );
        var func = cc.callFunc( this.exit.bind( this ) );
        this.runAction( cc.sequence( delay, func, null ) );
        this.fadeexit();

        if( this._resume ) this._resume.setEnabled( false );
        if( this._restart ) this._restart.setEnabled( false );
    },

    cb_sound: function( sender, enabled )
    {
        //TODO: soundEnabled
        EU.AudioEngine.soundEnabled(enabled);
        this.checkAudio();
    },

    cb_music: function( sender, enabled )
    {
        //TODO: soundEnabled
        EU.AudioEngine.soundEnabled(enabled);
        this.checkAudio();
    },

    checkAudio: function()
    {
        //TODO: soundEnabled
        var s = EU.AudioEngine.isSoundEnabled();
        //TODO: soundEnabled
        var m = EU.AudioEngine.isMusicEnabled();
        if( this._sound_off )this._sound_off.setVisible( !s );
        if( this._sound_on  )this._sound_on.setVisible( s );
        if( this._music_off )this._music_off.setVisible( !m );
        if( this._music_on )this._music_on.setVisible( m );
    },

    gameresume: function()
    {
        EU.AudioEngine.resumeAllEffects();
        this.setEnabled( false );
        this.removeFromParent();
    },

    restart: function()
    {
        this.setEnabled( false );
        this.removeFromParent();
        EU.GameGS.restartLevel();
    },

    exit: function()
    {
        this.setEnabled( false );
        cc.director.popScene();
        this.removeFromParent();
    },

    shop_did_closed: function()
    {
        this.setEnabled( true );
    },

    fadeexit: function()
    {
        var scene = cc.director.getRunningScene();
        var scores = scene.getChildByName("scorelayer");
        if(scores)
        {
            scores.removeFromParent();
        }

        this.runAction( cc.scaleTo( this.kFadeDuration, this._scaleFactor * 1.2 ) );
        this.runAction( cc.fadeTo( this.kFadeDuration, 0 ) );
    },

    fadeenter: function()
    {
        this.setScale( this._scaleFactor * 1.2 );
        this.runAction( cc.scaleTo( this.kFadeDuration, this._scaleFactor * 1.0 ) );
        this.runAction( cc.fadeTo( this.kFadeDuration, 255 ) );
    },

    checkFullscreen: function()
    {
        var fullscreen = EU.UserData.get_bool( "fullscreen", true );
        var item = this.getNodeByPath( this, "fullscreen" );
        if( item )
        {
            var on = item.getParamCollection().get( "on", "" );
            var off = item.getParamCollection().get( "off", "" );
            item.setImageNormal( fullscreen ? on : off );
        }
    }

});

EU.NodeExt.call(EU.GamePauseLayer.prototype);
